docs: Def NVIDIA driver and migration (#459)
NVIDIA/cloud-native-docs: gpu-operator/gpu-driver-configuration.rst

===============================================

> By default, the Helm chart configures a default NVIDIA driver custom resource during installation.

Reviewer: It configures default NVIDIA driver custom resource when `driver.nvidiaDriverCRD.deployDefaultCR` is set to `true`. We should update this line to mention that.

Reviewer: Ah nevermind, we do set deployDefaultCR to true by default. So that should get installed whenever nvidiadriverCRD is enabled. So we are good here.
https://github.com/NVIDIA/gpu-operator/blob/599e28368646116dc56a8018a068fbac6334cce1/deployments/gpu-operator/values.yaml#L141

A user-defined resource takes precedence on nodes that match its node selector, and the default resource
continues to manage the remaining GPU nodes.
User-defined resources must not select the same node.
If they do, the affected resources report ``notReady`` with a ``ConflictingNodeSelector`` condition,
and the Operator retains the existing driver ownership labels until you resolve the conflict.

Only one NVIDIA driver custom resource can have ``spec.default: true``.
A default resource cannot specify ``spec.nodeSelector`` because it would not be able to act as a fallback.
If more than one default resource exists, the Operator marks the affected resources ``notReady`` with a
``ReconcileFailed`` condition and does not change the existing driver ownership labels on nodes.
Delete or update the extra default resource to resume reconciliation.

To prevent the Helm chart from creating the default resource, specify the
``--set driver.nvidiaDriverCRD.deployDefaultCR=false`` argument when you install or upgrade the Operator.
Use this setting only if your user-defined resources select every GPU node that the Operator should manage,

> or if leaving some GPU nodes without an Operator-managed driver is intentional.

Reviewer: Or, a user wants to create their own fallback default nvidiadriver CR (by specifying `spec.default: true` for their nvidiadriver CR).

.. note::

>    A user-defined resource without a node selector matches all GPU nodes.

Reviewer: ```suggestion
   A user-defined NVIDIADriver custom resource without a node selector matches all GPU nodes.
```

Author: There's some precedent for dropping "custom" when the context is high that the subject is a custom resource.

Here's an example from upstream: https://kubernetes.io/docs/concepts/extend-kubernetes/operator/#example

Step 1 uses "A custom resource named SampleDB..." and step 4: "...find out what SampleDB resources are configured."

I anticipate your goal is precision.  My concern is that this level of repetition is harmful to human readers and likely inconsequent to agentic readers.

Reviewer: Ack, sounds good then.

>    The resource takes precedence over the default resource and conflicts with any other resource

Reviewer: ```suggestion
   The user defined NVIDIADriver CR takes precedence over the default CR and conflicts with any other CR
```

  Each NVIDIA driver custom resource can specify custom kernel module parameters by using a ConfigMap.
  For more information, refer to :doc:`Customizing NVIDIA GPU Driver Parameters during Installation <custom-driver-params>`.

.. _migrate-clusterpolicy-to-nvidiadriver:

***********************************************
Migrating from Cluster Policy Driver Management
***********************************************

You can migrate an existing Helm installation to NVIDIA driver custom resource management through
the controlled driver upgrade flow.
During the migration, the Operator assigns each GPU node to an NVIDIA driver custom resource and uses the
driver upgrade controller to replace the previous cluster policy managed driver pod on each node.

> When you migrate from a GPU Operator release earlier than v26.7.0, perform two Helm upgrades.

Reviewer: ```suggestion
When you migrate from a GPU Operator release earlier than v26.7.0, perform two Helm upgrades. Clusters on or after v26.7.0 don't need the two step upgrade and can switch from clusterpolicy to nvidiadriver in a single upgrade.
```

Author: Two things:

- First, I'd prefer to avoid introducing the two-upgrade process for < 26.7.0 and then immediately provide the counterpoint circumstance before a human reader even reads the rationale for the two-upgrade process.
- Second, the conditional in the procedure seems sufficiently clear: `#. If your current GPU Operator release is earlier than v26.7.0, upgrade...`.

My concern is that if we try to handle "< 26.7.0 and >= 26.7.0" sentence-by-sentence, that we're taxing the attention span of our human readers.

First, upgrade to v26.7.0 while retaining cluster policy driver management.

> Then, upgrade the same release again to enable NVIDIA driver custom resource management.

Reviewer: ```suggestion
Then, upgrade to the same release again enabling NVIDIADriver to enable NVIDIA driver custom resource management.
```

Author: I don't dispute that what you propose is technically accurate, but I find "...again enabling NVIDIADriver to enable NVIDIA driver..." difficult to read.

My proposal:

```
When you migrate from a GPU Operator release earlier than v26.7.0, perform two Helm upgrades:

* First, upgrade to v26.7.0 while retaining cluster policy driver management.
* Afterward, upgrade to the same release again.
  This time, enable NVIDIA driver custom resource management.

  The exact commands and arguments are shown in the following procedure.
```

Reviewer: Sure, sounds good to me.

This sequence starts the controller that supports controlled migration before changing driver ownership.

> Do not upgrade from an earlier release and enable NVIDIA driver custom resource management in the same

Reviewer: ```suggestion
Do not upgrade from an earlier release (using clusterpolicy for driver management) and enable NVIDIA driver custom resource management in the same
```

Author: My instinct is to nuke the whole "Do not..." para.  I think the instruction that precedes it and the procedure that follows are abundantly clear.

Reviewer: Sure, sounds good to me.

Helm operation because the old controller can remove the existing driver pods before the new controller
can take ownership.

Before you begin, verify the following requirements:

* The Helm values under ``driver`` represent the driver configuration that you want the chart-created
  default NVIDIA driver custom resource to use.
  The Operator does not copy changes made directly to the ``spec.driver`` field of the cluster policy
  custom resource into the new resource.
* Any existing non-default NVIDIA driver custom resources have non-overlapping node selectors.
  These resources become active when you enable NVIDIA driver custom resource management.
* ``driver.upgradePolicy.autoUpgrade`` is ``true`` so that the upgrade controller can perform the
  controlled replacement of the previous driver pods.
  Nodes with the ``nvidia.com/gpu-driver-upgrade.skip=true`` label are not migrated until you remove
  the label.
* Your workloads can tolerate the disruption configured by the driver upgrade policy.
  For more information, refer to :ref:`gpu-driver-upgrades`.

#. Identify the Helm release name:

   .. code-block:: console

      $ helm list -n gpu-operator

> #. If your current GPU Operator release is earlier than v26.7.0, upgrade to the target release while

Reviewer: ```suggestion
#. If your current GPU Operator release is earlier than v26.7.0 (and not using nvidiadriver CRs), upgrade to the target release while
```

Author: This whole procedure is under a "Migrating from Cluster Policy Driver Management" title, but I'll yield because we still have some human readers who sometimes do not take care.

  These resources become active when you enable NVIDIA driver custom resource management.
* ``driver.upgradePolicy.autoUpgrade`` is ``true`` so that the upgrade controller can perform the
  controlled replacement of the previous driver pods.
  Nodes with the ``nvidia.com/gpu-driver-upgrade.skip=true`` label are not migrated until you remove
  the label.
* Your workloads can tolerate the disruption configured by the driver upgrade policy.
  For more information, refer to :ref:`gpu-driver-upgrades`.

#. Identify the Helm release name:

   .. code-block:: console

      $ helm list -n gpu-operator

#. If your current GPU Operator release is earlier than v26.7.0 and the cluster does not already use
   ``NVIDIADriver`` resources, upgrade to the target release while
   retaining cluster policy driver management:

   .. code-block:: console

      $ helm upgrade <release-name> nvidia/gpu-operator \
          -n gpu-operator \
          --version=${version} \
          --reuse-values \
          --disable-openapi-validation \
          --set operator.upgradeCRD=true \
          --set driver.nvidiaDriverCRD.enabled=false \

>           --set driver.nvidiaDriverCRD.deployDefaultCR=false \

Reviewer: Do we need to set these (as they are the same as default value in chart)? We can avoid specifying them to avoid confusion where one might think these are always required.
